paddle/fluid/pybind/eager_method.cc

```diff
@@ -1243,6 +1242,10 @@ static PyObject* tensor_method__setitem_eager_tensor(TensorObject* self,
         value_tensor = egr::EagerAmpAutoCast(
             value_tensor.name(), value_tensor, amp_dtype, "set_value");
       }
+      if (value_tensor.initialized() &&
+          self->tensor.dtype() != value_tensor.dtype()) {
+        value_tensor = cast_ad_func(value_tensor, self->tensor.dtype());
```

Comment:
This is not a safe operants， check dtype first maybe

Reply:
Checkout the EagerAmpAutoCast && GetAmpDestDtype system

Reply:
self->tensor = set_value__dygraph_function(
          self->tensor, value_tensor, {}, {}, {}, attrs);